sktree/tree/unsupervised/_unsup_criterion.pyx

```diff
@@ -398,3 +397,175 @@ cdef class TwoMeans(UnsupervisedCriterion):
+            self.start,
+            self.end,
+            mean
+        ) / self.weighted_n_node_samples
+
+        impurity = n_node_samples*log(2*sig/mean)
+
+        return impurity
+
+    cdef void children_impurity(
+        self,
+        double* impurity_left,
+        double* impurity_right
+    ) nogil:
+        """Evaluate the impurity in children nodes.
+
+        i.e. the impurity of the left child (sample_indices[start:pos]) and the
+        impurity the right child (sample_indices[pos:end]).
+
+        TODO: describe FastBIC
+
+        [Equation here]
+
+        - left_variance = left_weight * left_impurity / n_sample_of_left_child
+        - right_variance = right_weight * right_impurity / n_sample_of_right_child
+
+        Parameters
+        ----------
+        impurity_left : double pointer
+            The memory address to save the impurity of the left node
+        impurity_right : double pointer
+            The memory address to save the impurity of the right node
+        """
+        cdef SIZE_t pos = self.pos
+        cdef SIZE_t start = self.start
+        cdef SIZE_t end = self.end
+        cdef SIZE_t s_l
+        cdef SIZE_t s_r
+        cdef double p_l
+        cdef double p_r
+        cdef double mean_left
+        cdef double mean_right
+        cdef double sig_left
+        cdef double sig_right
+        cdef double left_term
+        cdef double right_term
+
+        # number of samples of left and right
+        s_l = pos - start
+        s_r = end - pos
+
+        # compute prior
+        p_l = s_l / self.n_node_samples
+        p_r = s_r / self.n_node_samples
+
+        # first compute mean of left and right
+        mean_left = self.sum_left / self.weighted_n_left
+        mean_right = self.sum_right / self.weighted_n_right
+
+        sig_left = self.sum_of_squares(
+            start,
+            pos,
+            mean_left
+        ) / self.weighted_n_left
+
+        sig_right = self.sum_of_squares(
+            pos,
+            end,
+            mean_right
+        ) / self.weighted_n_right
+
+        left_term = log(2*p_l*sig_left/mean_left)
+        right_term = log(2*p_r*sig_right/mean_right)
+
+        # set values at the address pointer
+        impurity_left[0] = s_l*left_term + s_r*right_term
+        impurity_right[0] = s_r*left_term + s_l*right_term
+
+    cdef void set_sample_pointers(
+        self,
+        SIZE_t start,
+        SIZE_t end
+    ) nogil:
+        """Set sample pointers in the criterion.
+
+        Set given start and end sample_indices. Also will update node statistics,
+        such as the `sum_total`, which tracks the total value within the current
+        node for sample_indices[start:end].
+
+        Parameters
+        ----------
+        start : SIZE_t
+            The start sample pointer.
+        end : SIZE_t
+            The end sample pointer.
+        """
+        self.n_node_samples = end - start
+        self.start = start
+        self.end = end
```

Comment:
Lets move this to the base class and also delete the corresponding function inside `TwoMeans`